bip-ccv.mediawiki

```diff
@@ -0,0 +1,420 @@
+::'''Figure 2:''' Three UTXOs aggregate their amounts towards the same output contract, using <code>CCV</code> with the <i>default</i> semantic.
+
+-----
+
+::[[File:bip-ccv/amount_example_3.png|framed|center|alt=split amount logic|600px]]
+::'''Figure 3:''' A UTXO <code>A</code> sends a portion of its amount to a contract <code>A'</code> identical to itself, and the rest to a different contract <code>B</code>. It would use <code>CCV</code> to introspect its own input's program, then to check the first output with the <i>deduct</i> semantic, then to check the second output with the <i>default</i> semantic to assign the residual amount.
+
+-----
+
+::[[File:bip-ccv/amount_example_4.png|framed|center|alt=split and aggregate amount logic|600px]]
+::'''Figure 4:''' Similar to the previous example, but a second input <code>B</code> also checks the same output <code>X</code> with the <i>default</i> semantic, aggregating its input with the residual amount of the first input.
+
+-----
+
+Note that the ''deduct'' semantic does not allow to check the exact amount of its output. Therefore, in contracts using a scheme similar to figure 3 or 4 above, amounts be constrained with a signature, or with future introspection opcodes that allow fixing the amount.
+
+== Specification ==
+
+The tapscript opcode <code>OP_SUCCESS187</code> (<code>0xbb</code>) is constrained with new rules to implement <code>OP_CHECKCONTRACTVERIFY</code>.
+
+When evaluating <code>OP_CHECKCONTRACTVERIFY</code> (<code>OP_SUCCESS187</code>,
+<code>0xbb</code>), the expected format of the stack, shown top to bottom, is:
+
+<source>
+<flags>
+<taptree>
+<pk>
+<index>
+<data>
+</source>
+
+where:
+
+* <code><flags></code> is a minimally encoded integer, according to one of the values defined below.
+* <code><taptree></code> is the Merkle root of the taproot tree, or a minimally encoded <code>-1</code>, or the empty buffer.
+* <code><pk></code> is called the ''naked key'', and it's a valid 32-byte x-only public key, or a minimally encoded <code>-1</code>, or the empty buffer.
+* <code><index></code> is a minimally encoded -1, or a minimally encoded non-negative integer.
+* <code><data></code> is a buffer of arbitrary length.
+
+In short, the semantics of the opcode with respect to the Script can be summarized as follows:
+
+<blockquote>
+Verify that the input/output with the given 'index' is a P2TR UTXO whose taproot output key is obtained from 'pk', tweaked with the hash of 'data' (if non-empty), then taptweaked with 'taptree' (if non-empty).
+</blockquote>
+
+If the <code><data></code> is non-empty, then the additive tweak for the data is computed as:
+
+<source lang="python">
+  data_tweak = sha256(pk || data)
+</source>
+
+In the following, the ''current input'' is the input whose Script is being executed.
+
+The following value of the <code><flags></code> are defined:
+* <code>CCV_FLAG_CHECK_INPUT = -1</code>: Check an input's script; no amount check.
```

Comment:
Most flags in Bitcoin Core are usigned integers, which seems easier to read. But I guess this is because the script interpreter uses (variable size) signed integers? On the bright side, it means you can expand the number of bit flags easily...

When this flag is absent, I assume you don't check the input?

Reply:
The flag can't be 'absent', as it's the value of the `flags` parameter. Note that it's not a bitmap: -1 means "check input", 0, 1 and 2 are the three defferent _amount_ behaviors for checking the output (default, deduct and ignore).

Perhaps the confusion comes from calling it `flags` - let me know if you have suggestions for a better name.

Reply:
At least to me flags imply a bitmap. Terminology aside, why not use a bitmap?

Since they're all mututally exclusive, I would probably call it "mode".